File: wpf_sample_net/entities/Address.cs
using System;
using System.Collections.Generic;

using wpf_lib.lib;
using wpf_lib.lib.attributes;

using wpf_generated.functions;

namespace wpf_generated.entities {


  public class Address : EntityBase {

    // Regular Attributes
    private string _unitNumber;
    public string UnitNumber {
      get { return _unitNumber; }
      set {
        _unitNumber = value;
        RaisePropertyChanged(nameof(UnitNumber));

        RaisePropertyChanged(nameof(FirstAddressLine));
      }
    }
    private string _theAddress;
    public string TheAddress {
      get { return _theAddress; }
      set {
        _theAddress = value;
        RaisePropertyChanged(nameof(TheAddress));

        RaisePropertyChanged(nameof(FirstAddressLine));
      }
    }
    private string _city;
    public string City {
      get { return _city; }
      set {
        _city = value;
        RaisePropertyChanged(nameof(City));

        RaisePropertyChanged(nameof(SecondAddressLine));
      }
    }
    private string _stateOrProvince;
    public string StateOrProvince {
      get { return _stateOrProvince; }
      set {
        _stateOrProvince = value;
        RaisePropertyChanged(nameof(StateOrProvince));

        RaisePropertyChanged(nameof(SecondAddressLine));
      }
    }
    private string _zip;
    public string Zip {
      get { return _zip; }
      set {
        _zip = value;
        RaisePropertyChanged(nameof(Zip));

        RaisePropertyChanged(nameof(ThirdAddressLine));
      }
    }

    // Derived Attributes
    public string FirstAddressLine {
      get {
        return TheAddress + "   Unit " + UnitNumber;
      }
    }
    public string SecondAddressLine {
      get {
        return City + ", " + StateOrProvince;
      }
    }
    public string ThirdAddressLine {
      get {
        return Zip;
      }
    }

    // Associations

    // Validations
    public override void CalculateErrors(string prefix, EntityErrors errors) {
      if (string.IsNullOrWhiteSpace(TheAddress?.ToString()))
        errors.Add("Name is required", prefix, nameof(TheAddress));
    }

    public static Address Create(EntityBase owner) {
      return new Address {
        Owner = owner,
      };
    }
  }
}

File: platforms/ts-chakra/src/x10_generated/small/entities/Address.ts
import { v4 as uuid } from 'uuid';

import { addError, type FormError } from 'react_lib/form/FormProvider';
import isBlank from 'react_lib/utils/isBlank';
import x10toString from 'react_lib/utils/x10toString';

import { AppContextType } from 'SmallAppContext';
import { type Country } from 'x10_generated/small/entities/Country';


// Type Definition
export type Address = {
  readonly id: string,
  readonly unitNumber?: string,
  readonly theAddress?: string,
  readonly city?: string,
  readonly stateOrProvince?: string,
  readonly zip?: string,
  readonly country?: Country,
};


// Derived Attribute Functions
export function addressFirstAddressLine(appContext: AppContextType, address?: {
    theAddress?: string,
    unitNumber?: string,
} | null | undefined): string | undefined {
  if (address == null) return '';
  const result = x10toString(address?.theAddress) + '   Unit ' + x10toString(address?.unitNumber);
  return result;
}

export function addressSecondAddressLine(appContext: AppContextType, address?: {
    city?: string,
    stateOrProvince?: string,
} | null | undefined): string | undefined {
  if (address == null) return '';
  const result = x10toString(address?.city) + ', ' + x10toString(address?.stateOrProvince);
  return result;
}

export function addressThirdAddressLine(appContext: AppContextType, address?: {
    zip?: string,
} | null | undefined): string | undefined {
  if (address == null) return '';
  const result = address?.zip;
  return result;
}



// Create Default Function
export function createDefaultAddress(): Address {
  return {
    id: uuid(),
    unitNumber: '',
    theAddress: '',
    city: '',
    stateOrProvince: '',
    zip: '',
    country: undefined,
  };
}


// Validations
export function addressCalculateErrors(appContext: AppContextType, address?: Address, prefix?: string, inListIndex?: number): FormError[] {
  const errors: FormError[] = [];
  if (address == null ) return errors;

  if (isBlank(address.theAddress))
    addError(errors, 'The Address is required', ['theAddress'], prefix, inListIndex);
  if (isBlank(address.city))
    addError(errors, 'City is required', ['city'], prefix, inListIndex);
  if (isBlank(address.stateOrProvince))
    addError(errors, 'State Or Province is required', ['stateOrProvince'], prefix, inListIndex);
  if (isBlank(address.zip))
    addError(errors, 'Zip is required', ['zip'], prefix, inListIndex);
  if (isBlank(address.country))
    addError(errors, 'Country is required', ['country'], prefix, inListIndex);

  return errors;
}

